### DISCUSSION TITLE
Update system list item fields

### DISCUSSION THREAD

**Discussion by Nevinia210** [pnp/pnpcore#672] [Q&A] [answered]
Hello Everyone,
is there any way how to update system list item fields:

Created
Created By
Modified
Modified By

on Site Pages?
I'm using C# code

Thank you in advance for your answers

**Comment by jansenbe** [comment, marked as answer]
Hi @Nevinia210 , thanks for using PnP Core SDK.

I've just added documentation (https://pnp.github.io/pnpcore/using-the-sdk/pages-intro.html#updating-the-page-author-editor-created-and-modified-system-properties) showing how this is done for a page. This is quite similar to how this is done for a regular list item (see https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#updating-the-list-item-author-editor-created-and-modified-system-properties). There are also sample tests: https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/PagesTests.cs#L2003 and https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/ListItemTests.cs#L380

**Reply by Nevinia210** [reply]
Hi @jansenbe ,
thank you for updating documentation, i tried your approach and end up with an error:

<pre lang="C#"><code>
var page = context.Web.GetPages(pageName).FirstOrDefault();
var pageListItem = page.PageListItem;

var siteUsers = context.Web.SiteUsers.ToList();
var selectedUser = siteUsers.FirstOrDefault(x => x.Id == myUserId);

IField myFieldAuthor = targetSitePagesLibrary.Fields.FirstOrDefault(x => x.InternalName == "Author");
IField myFieldEditor = targetSitePagesLibrary.Fields.FirstOrDefault(x => x.InternalName == "Editor");

var singleUserValueAuthor = myFieldAuthor.NewFieldUserValue(selectedUser);
var singleUserValueEditor = myFieldEditor.NewFieldUserValue(selectedUser);

pageListItem["Author"] = singleUserValueAuthor;
pageListItem["Editor"] = singleUserValueEditor;

pageListItem.UpdateOverwriteVersion();
</code></pre>

Error message: "CSOM service exception"
Detailed message: "The 'Type' attribute is invalid - The value 'IFile' is invalid according to its datatype 'http://schemas.microsoft.com/sharepoint/clientquery/2009:MethodParameterTypeType' - The Enumeration constraint failed."

Am I missing something?
Also is it possible to update Author and Editor fields to different user than current one?

**Reply by jansenbe** [reply]
@Nevinia210 : the example shows how to load the pageFile (`var pageFile = await myPage.GetPageFileAsync(p => p.ListItemAllFields);`) which you've left out. It's best to get a fresh page list item in that manner. Try updating your sample like this:

```csharp
var page = context.Web.GetPages(pageName).FirstOrDefault();
var pageFile = page.GetPageFile(p => p.ListItemAllFields);
var pageListItem = pageFile.ListItemAllFields;

// rest of your code
```

If I'm not mistaken these system updates are only allowed if all 4 fields are updated, so if with my above recommendation things still don't work then also try setting Created and Modified.

Please report back, if needed we'll adjust the docs with your learnings.

**Reply by Nevinia210** [reply]
@jansenbe : I updated code accordingly, now I didn't receive any error, but also no changes are happened in List Item fields.
I also tried .SystemUpdate(), but that didn't do the trick either

<pre lang="C#"><code>
var page = context.Web.GetPages(pageName).FirstOrDefault();
var pageListItem = page.GetPageFile(p => p.ListItemAllFields).ListItemAllFields;

var siteUsers = context.Web.SiteUsers.ToList();
var selectedUser = siteUsers.FirstOrDefault(x => x.Id == myUserId);

IField myFieldAuthor = targetSitePagesLibrary.Fields.FirstOrDefault(x => x.InternalName == "Author");
IField myFieldEditor = targetSitePagesLibrary.Fields.FirstOrDefault(x => x.InternalName == "Editor");

var singleUserValueAuthor = myFieldAuthor.NewFieldUserValue(selectedUser);
var singleUserValueEditor = myFieldEditor.NewFieldUserValue(selectedUser);

pageListItem["Author"] = singleUserValueAuthor;
pageListItem["Editor"] = singleUserValueEditor;
pageListItem["Created"] = new DateTime(2020, 6, 6);
pageListItem["Modified"] = new DateTime(2020, 6, 6);

pageListItem.UpdateOverwriteVersion();
</code></pre>

**Reply by jansenbe** [reply]
Hmm, you're setting another user account than the one you're running the code with?

**Reply by Nevinia210** [reply]
yes, user that i set will change list item from list item

**Reply by jansenbe** [reply]
Below works for me, setting the user and date correct. 

```csharp
  var newPage = await context.Web.NewPageAsync();
  string pageName ="demo.aspx";

  newPage.AddSection(CanvasSectionTemplate.ThreeColumn, 1, VariantThemeType.Soft, VariantThemeType.Strong);
  newPage.AddControl(newPage.NewTextPart("I"), newPage.Sections[0].Columns[0]);
  newPage.AddControl(newPage.NewTextPart("like"), newPage.Sections[0].Columns[1]);
  newPage.AddControl(newPage.NewTextPart("PnP"), newPage.Sections[0].Columns[2]);

  // Save the page
  await newPage.SaveAsync(pageName);

  // Load the Page File
  var pageFile = await newPage.GetPageFileAsync(p => p.ListItemAllFields);

  // Ensure the user to set does exist on the site
  var currentUser = await context.Web.EnsureUserAsync("kevinc@bertonline.onmicrosoft.com");
  var newDate = new DateTime(2020, 10, 20);

  // Load the Author and Editor fields                
  var author = newPage.PagesLibrary.Fields.AsRequested().FirstOrDefault(p => p.InternalName == "Author");
  var editor = newPage.PagesLibrary.Fields.AsRequested().FirstOrDefault(p => p.InternalName == "Editor");

  pageFile.ListItemAllFields["Author"] = author.NewFieldUserValue(currentUser);
  pageFile.ListItemAllFields["Editor"] = editor.NewFieldUserValue(currentUser);
  pageFile.ListItemAllFields["Created"] = newDate;
  pageFile.ListItemAllFields["Modified"] = newDate;
  await pageFile.ListItemAllFields.UpdateOverwriteVersionAsync();
```

**Reply by Nevinia210** [reply]
Hi @jansenbe,

I finally found where was the issue:
Sites.FullControl.All permission for my app is needed to do that,
Previously I had only Sites.Manage.All and that is not enough in case of setting system properties

Thank you for your help

**Reply by jansenbe** [reply]
I'll update the docs to clarify that requirement, thanks for the feedback.
